Source-build cli/apphost/static/singlefilehost native build failure: "pal_ssl.c:710: undefined reference to `local_X509_new'"

In https://github.com/dotnet/source-build/pull/1782, we're trying to update source-build to 5.0.0-rc.1, but hitting an issue in the installer native build with cli/apphost/static/singlefilehost, during a non-portable build (link to system libraries):

```
    [100%] Built target hostpolicy
    /work/artifacts/src/runtime.38017c3935de95d0335bac04f4901ddfc2718656/artifacts/bin/native/net5.0-Linux-Release-x64/libSystem.Security.Cryptography.Native.OpenSsl.a(pal_ssl.c.o): In function `CryptoNative_OpenSslGetProtocolSupport':
    /work/artifacts/src/runtime.38017c3935de95d0335bac04f4901ddfc2718656/src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c:710: undefined reference to `local_X509_new'
    clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
    gmake[2]: *** [cli/apphost/static/singlefilehost] Error 1
    gmake[1]: *** [cli/apphost/static/CMakeFiles/singlefilehost.dir/all] Error 2
    gmake: *** [all] Error 2
    Errors
        /work/artifacts/src/runtime.38017c3935de95d0335bac04f4901ddfc2718656/src/installer/corehost/corehost.proj(46,5): error MSB3073: The command "/work/artifacts/src/runtime.38017c3935de95d0335bac04f4901ddfc2718656/src/installer/corehost/build.sh Release x64 -apphostver "5.0.0-rc.1.20451.14" -hostver "5.0.0-rc.1.20451.14" -fxrver "5.0.0-rc.1.20451.14" -policyver "5.0.0-rc.1.20451.14" -commithash "78f189bca6ae9a9d1a994753eae9f88d59b80ef9" -os Linux -cmakeargs "-DVERSION_FILE_PATH=/work/artifacts/src/runtime.38017c3935de95d0335bac04f4901ddfc2718656/artifacts/obj/_version.c" -portablebuild=false -keepnativesymbols -coreclrartifacts /work/artifacts/src/runtime.38017c3935de95d0335bac04f4901ddfc2718656/artifacts/bin/coreclr/Linux.x64.Release/ -nativelibsartifacts /work/artifacts/src/runtime.38017c3935de95d0335bac04f4901ddfc2718656/artifacts/bin/native/net5.0-Linux-Release-x64" exited with code 2. [/work/artifacts/src/runtime.38017c3935de95d0335bac04f4901ddfc2718656/src/installer/corehost/corehost.proj]
```

This seems to be around the same area as we had issues with in https://github.com/dotnet/runtime/issues/41768, but I'm not good enough at cmake builds to really know what to look at. Here are the same set of logs as last time plus a binlog though, from a build I did on my own machine with a CentOS 7 Docker image:

[5.0-rc1-host-logs.zip](https://github.com/dotnet/runtime/files/5320334/5.0-rc1-host-logs.zip)

@omajid @janvorli can you help diagnose this?

(Manually labeling based on this being a native build error in the host, feel free to relabel.)

--- a/src/libraries/Native/Unix/System.Security.Cryptography.Native/opensslshim.h
+++ b/src/libraries/Native/Unix/System.Security.Cryptography.Native/opensslshim.h
@@ -1092,7 +1092,6 @@ FOR_ALL_OPENSSL_FUNCTIONS
 #define SSL_is_init_finished local_SSL_is_init_finished
 #define X509_CRL_get0_nextUpdate local_X509_CRL_get0_nextUpdate
 #define X509_NAME_get0_der local_X509_NAME_get0_der
-#define X509_new local_X509_new
 #define X509_PUBKEY_get0_param local_X509_PUBKEY_get0_param
 #define X509_STORE_CTX_get0_cert local_X509_STORE_CTX_get0_cert
 #define X509_STORE_CTX_get0_chain local_X509_STORE_CTX_get0_chain

